### ISSUE TITLE
[Bug]: Live Reload not working.

### ISSUE THREAD

**Issue by drewkill32** [remix-run/remix#1601]
### Which Remix packages are impacted?

- `remix` (Remix core)
- `@remix-run/react`
- `@remix-run/serve`

### What version of Remix are you using?

1.1.3

### What version of Node are you using? Minimum supported version is 14.

14.17.4

### Steps to Reproduce

1. starting from the [Quickstart](https://remix.run/docs/en/v1/tutorials/blog) template. 
2. create a posts route. 
3. navigate to route
4. the route does not show up

Every step along the way I have to restart the dev server to see the changes in the web. For example I added 

```jsx

export default function Posts() {
    const posts = useLoaderData();
    console.log('test3');
    return (
        <div>
            <h1>Posts</h1>
        </div>
    );
}
```

and the console still shows that it reloaded but still shows test2 (the value from when the dev server started)

Navigated to http://10.202.57.104:3000/posts
index.tsx:18 test2
posts:23 💿 File changed: app\routes\posts\index.tsx
posts:23 💿 Rebuilding...
posts:23 💿 Rebuilt in 201ms
posts:26 💿 Reloading window ...
Navigated to http://10.202.57.104:3000/posts
index.tsx:18 test2

I'm on Windows using  Chrome(97.0.4692.71) and Edge(Version 97.0.1072.62), both failed 

### Expected Behavior

the hot reload should show the changes without having to restart

### Actual Behavior

I have to stop the dev server and restart to see any changes

**Comment by jacob-ebey** [comment]
Check that the <LiveReload /> component is making it onto your page. If it's not, the first thing to check is what the value of NODE_ENV is when you run. If it's not "development", set it before you start the dev command.

**Comment by drewkill32** [comment]
The component is on the page. I just double-checked the NODE_ENV by adding `<p>{process.env.NODE_ENV}</p>` to the component and restarted the server. It is in development

**Comment by drewkill32** [comment]
I still have not found a solution other than it the hot-reloading it working on my Linux machine but refuses to work on Windows.

**Comment by kurdi-dev** [comment]
I'm also using Windows and I realized hot-reloads only work when I make changes inside the app directory of my Remix project. 
I don't know if this is how it is supposed to work or it's a bug!

**Comment by adamhp** [comment]
> I'm also using Windows and I realized hot-reloads only work when I make changes inside the app directory of my Remix project. I don't know if this is how it is supposed to work or it's a bug!

Interesting observation. I was struggling with this and your point made me realize I currently have all my components in a  `components/` folder outside of my `app/` directory. I just did a brief experiment and moving a component into the `app/` directory seems to fix it — i.e. causes changes to reflect on the hot reload live.

As a possible feature, could we provide paths to whatever is watching the filesystem for hot reload?

**Comment by seyaobey-dev** [comment]
In my case I have all my components inside the `./app` folder. But hot-reload is still not work. I am on MacOS

**Comment by tomsutton1984** [comment]
> I still have not found a solution other than it the hot-reloading it working on my Linux machine but refuses to work on Windows.

I hit the same problems, then realised that WSL was sat in the middle. Live Reload started working perfectly once I moved my working directory to the Linux filesystem. https://docs.microsoft.com/en-us/windows/wsl/setup/environment#file-storage

**Comment by nippur72** [comment]
same problem here on Windows and `create-remix`, I'm trying to follow the QuickStart tutorial but it doesn't reload 😢

**Comment by Tymek** [comment]
> Live Reload started working perfectly once I moved my working directory to the Linux filesystem. https://docs.microsoft.com/en-us/windows/wsl/setup/environment#file-storage

It doesn't seem to work on my end, even though I have project files in `\\wsl$\...`

**Comment by machour** [comment]
@Tymek the warning is unrelated, check #1077

**Comment by Tymek** [comment]
Thx.

FIY everybody, if you move to Remix Stack (Blues) there is no issue with hot reloading. It is using pm2 server internally for that. That's what I did.

**Comment by mayank99** [comment]
Hot reload doesn't work for me on `create-remix@1.3.3` with the starter templates. I'm on Windows with WSL2.

I dug a little deeper and found the root cause here: https://github.com/microsoft/WSL/issues/4739

This issue comes up across many tools, and the workaround is often to switch to polling.
- For example, `chokidar` provides the `usePolling` flag: https://github.com/paulmillr/chokidar#performance
- Vite also forwards the `usePolling` flag from `chokidar` for this exact reason: https://vitejs.dev/config/#server-watch
- `esbuild` watch mode uses polling by default: https://esbuild.github.io/api/#watch

Maybe remix should also offer a polling option?

**Comment by mikestopcontinues** [comment]
I don't mean to +1, but I'm also getting the issue on mac, with a fresh vercel/ts boilerplate.

**Comment by Devr-pro** [comment]
Same issue with me 
Am on Windows10 and all setup of remix was good but I then realized that My localhost isnt hotreloading on changes..,
I have used this exp   `     {process.env.NODE_ENV === 'development' ? <LiveReload/> : null}`  ..,But still my issue is persisting.

**Comment by kiliman** [comment]
They had started removing `cross-env NODE_ENV=development` from the *package.json* scripts. Due to the order in which things were occurring, the internal `<LiveReload/>` component wasn't getting the property environment.

Re-add the cross-env and it should work.

**Comment by mikestopcontinues** [comment]
@kiliman It doesn't work with or without manually setting the `NODE_ENV` via CLI, `.env`, or in the config files. `NODE_ENV` is 'development' without any of them anyway.

**Comment by BiffBish** [comment]
Any known fixes? Running remix on WSL Ubuntu 20.04

**Comment by BiffBish** [comment]
> `<LiveReload/>` is a simple component. You can simply inline it in your _root_ route.
> 
> ```ts
> export const LiveReload =
>   process.env.NODE_ENV !== "development"
>     ? () => null
>     : function LiveReload({
>         port = Number(process.env.REMIX_DEV_SERVER_WS_PORT || 8002),
>       }: {
>         port?: number;
>       }) {
>         let setupLiveReload = ((port: number) => {
>           let protocol = location.protocol === "https:" ? "wss:" : "ws:";
>           let host = location.hostname;
>           let socketPath = `${protocol}//${host}:${port}/socket`;
> 
>           let ws = new WebSocket(socketPath);
>           ws.onmessage = (message) => {
>             let event = JSON.parse(message.data);
>             if (event.type === "LOG") {
>               console.log(event.message);
>             }
>             if (event.type === "RELOAD") {
>               console.log("💿 Reloading window ...");
>               window.location.reload();
>             }
>           };
>           ws.onerror = (error) => {
>             console.log("Remix dev asset server web socket error:");
>             console.error(error);
>           };
>         }).toString();
> 
>         return (
>           <script
>             suppressHydrationWarning
>             dangerouslySetInnerHTML={{
>               __html: `(${setupLiveReload})(${JSON.stringify(port)})`,
>             }}
>           />
>         );
>       };
> ```

This doesn't seem to work? i start up the server and when i make a change to a document and do `ctrl+s` nothing happens int he console and the page doesn't recompile and build
![image](https://user-images.githubusercontent.com/42410516/160941208-e5cabe51-0b99-40d3-91da-d73900defe80.png)

**Comment by nilobarp** [comment]
Had the same issue. Fixed by changing `remix.confix.js`
**from**:
```js
module.exports = {
  ignoredRouteFiles: [".*"],
};
```
**to**:
```js
module.exports = {
  cacheDirectory: "./node_modules/.cache/remix",
  ignoredRouteFiles: [".*", "**/*.css", "**/*.test.{js,jsx,ts,tsx}"],
};
```

**Comment by mikestopcontinues** [comment]
@nilobarp That's already in my `remix.config.js`. Why would it matter either way? My best guess is these solutions are all part of an elaborate April fools prank. 😁

**Comment by kiliman** [comment]
I believe there was a bug in the dev tool in that it wasn't passing the `REMIX_DEV_SERVER_WS_PORT` variable to the compiler. Have you tried the latest version? Remix now gets a random unused port for the WS port.

**Comment by nilobarp** [comment]
@mikeybinnswebdesign wish it was a prank. I tried everything mentioned in this thread, the only thing that worked was the change in `remix.config.js`. I undid the change reload stopped working, bring it back reload works again.

This is too much magic for my brain.

**Comment by mikestopcontinues** [comment]
@nilobarp @kiliman — Just figured out, for me it was disabling `devServerPort` which doesn't seem to be working anyway. Thanks!

**Comment by LukeXF** [comment]
LiveReload does not work for me,

I copied the code that @kiliman shared and added more console logs. The frontend never seems to connect to the WS server and is just constantly pending.


<img width="324" alt="image" src="https://user-images.githubusercontent.com/6208873/162259220-7728fe18-c785-46b8-9183-7dd90457fadb.png">

No `ws.onmessage` console.log is ever ran either.

Any suggestions?

**Comment by kiliman** [comment]
@LukeXF Are you on Mac or Windows?

Did you add/change `devServerPort` in remix.config?

Also check to see if `REMIX_DEV_SERVER_WS_PORT` environment variable is defined.

Worst case, just update your `dev` script to:

`"cross-env NODE_ENV=development PORT=3000 REMIX_DEV_SERVER_WS_PORT=8002 remix dev"`

**Comment by machour** [comment]
@drewkill32 is this issue fixed for you ?

**Comment by drewkill32** [comment]
Just tried again with the latest from remix. No it hasn't. If I run it on my Linux machine or Linux Subsystem for Windows the hot reload works.

**Comment by LukeXF** [comment]
I tried adding `"cross-env NODE_ENV=development PORT=3000 REMIX_DEV_SERVER_WS_PORT=8002 remix dev"` but I get:

`Error: remix dev is not supported for custom servers.` Because I'm using CF Pages, anyway I revmoed the custom server: "./server.js" code and I get `error: ReferenceError: crypto is not defined`


```
{
  error: ReferenceError: crypto is not defined
  at createKey (/Users/lukebrown/Projects/example/remix/node_modules/@remix-run/cloudflare/crypto.js:35:13)
  at unsign (/Users/lukebrown/Projects/example/remix/node_modules/@remix-run/cloudflare/crypto.js:27:19)
  at decodeCookieValue (/Users/lukebrown/Projects/example/remix/node_modules/@remix-run/server-runtime/cookies.js:87:33)
  at Object.parse (/Users/lukebrown/Projects/example/remix/node_modules/@remix-run/server-runtime/cookies.js:53:66)
  at Object.getSession (/Users/lukebrown/Projects/example/remix/node_modules/@remix-run/server-runtime/sessions/cookieStorage.js:36:67)
  at loader (/Users/lukebrown/Projects/example/remix/app/root.tsx:145:39)
  at Object.callRouteLoader (/Users/lukebrown/Projects/example/remix/node_modules/@remix-run/server-runtime/data.js:77:20)
  at /Users/lukebrown/Projects/example/remix/node_modules/@remix-run/server-runtime/server.js:259:113
  at Array.map (<anonymous>)
  at handleDocumentRequest (/Users/lukebrown/Projects/example/remix/node_modules/@remix-run/server-runtime/server.js:259:67)
}
```

Then I get `It appears you're using a module that is built in to node, but you installed it as a dependency which could cause problems. Please remove crypto before continuing.` very odd! Didn't think I'd need to install crypto.

Anyway, I'm using v16.13.0 on a Mac M1 Chip

REMIX_DEV_SERVER_WS_PORT is not used, but I have tried it, doesn't change anything, same with devServerPort, not using either at the moment.

http://localhost:8002/ still returns 404 

I will investigate further on a clean project!

**Comment by kiliman** [comment]
> http://localhost:8002/ still returns 404

Odd. If the web socket server wasn't running at all, you should have gotten a "can't connect error".

Have you tried rebooting to make sure there isn't some stray process hanging onto that port?

**Comment by michael-marid** [comment]
If you are using vs code (windows), you could do the following:

1.  Install Remote - WSL extension
2.  Open the green connection button (down-left) to connect to the distro (ex. Ubuntu).
3.  Git clone your project there and use the docker-compose file or run the server.

**Comment by eran-or** [comment]
Interesting behavior: When I clone the demo with npx create-remix --template remix-run/indie-stack blog-tutorial
on my local machine the hot reloading works fine but when I use Gitpod then it's not working. so I guess it is something that relate to this (I'm working with mac 2013, in case it's matter)

**Comment by ahmed1107** [comment]
> LiveReload does not work for me,
> 
> I copied the code that @kiliman shared and added more console logs. The frontend never seems to connect to the WS server and is just constantly pending.
> 
> <img alt="image" width="324" src="https://user-images.githubusercontent.com/6208873/162259220-7728fe18-c785-46b8-9183-7dd90457fadb.png">
> 
> No `ws.onmessage` console.log is ever ran either.
> 
> Any suggestions?

Im experiancing the same using windows server

**Comment by drewkill32** [comment]
> 

When I create a new project from the indie-stack on a new Windows machine I still don't have hot reloading working. I can only get it to work using WSL

**Comment by intcreator** [comment]
to clarify the workaround that some other people have suggested, the Remix project needs to be installed in the WSL only portion of the WSL filesystem. it may look like `/mnt/c/Users/[username]/projects/my-remix-app` is in the WSL filesystem but since it's an abstraction of the Windows filesystem it will not work. however, it should work fine if you set up the project in `/home/[username]/my-remix-app` for example. the reason this works is the WSL filesystem is able to detect file changes in its own virtualized environment but not in its abstraction of the Windows filesystem

this workaround requires setting up your code editor to point to the WSL files which isn't an option for everyone, so I support an option for polling to be added as a short term solution while WSL itself gets the bug fixed

**Comment by gonghaibo** [comment]
I don't want to chat here or anything. But I ran into this problem.
And I still don't know how to solve it.

So how to solve the macbook pro(m1) 11.6

**Comment by RATIU5** [comment]
This issue still occurs on Windows with Remix 1.4.1. Creating or deleting files outside of `app/routes` works fine and live reload does not fail. Creating or deleting any type of file inside the `routes` folder reloads once then refuses to reload upon any further changes.

**Comment by amanolar94** [comment]
> `<LiveReload/>` is a simple component. You can simply inline it in your _root_ route.
> 
> ```ts
> export const LiveReload =
>   process.env.NODE_ENV !== "development"
>     ? () => null
>     : function LiveReload({
>         port = Number(process.env.REMIX_DEV_SERVER_WS_PORT || 8002),
>       }: {
>         port?: number;
>       }) {
>         let setupLiveReload = ((port: number) => {
>           let protocol = location.protocol === "https:" ? "wss:" : "ws:";
>           let host = location.hostname;
>           let socketPath = `${protocol}//${host}:${port}/socket`;
> 
>           let ws = new WebSocket(socketPath);
>           ws.onmessage = (message) => {
>             let event = JSON.parse(message.data);
>             if (event.type === "LOG") {
>               console.log(event.message);
>             }
>             if (event.type === "RELOAD") {
>               console.log("💿 Reloading window ...");
>               window.location.reload();
>             }
>           };
>           ws.onerror = (error) => {
>             console.log("Remix dev asset server web socket error:");
>             console.error(error);
>           };
>         }).toString();
> 
>         return (
>           <script
>             suppressHydrationWarning
>             dangerouslySetInnerHTML={{
>               __html: `(${setupLiveReload})(${JSON.stringify(port)})`,
>             }}
>           />
>         );
>       };
> ```

This is the only solution that worked for me

**Comment by haach** [comment]
> Had the same issue. Fixed by changing `remix.confix.js` **from**:
> 
> ```js
> module.exports = {
>   ignoredRouteFiles: [".*"],
> };
> ```
> 
> **to**:
> 
> ```js
> module.exports = {
>   cacheDirectory: "./node_modules/.cache/remix",
>   ignoredRouteFiles: [".*", "**/*.css", "**/*.test.{js,jsx,ts,tsx}"],
> };
> ```

This worked for me! Thanks @nilobarp After a server restart that reload was hot hot hot 🔥

**Comment by connordear** [comment]
Experience on Windows: I was running `remix dev` through a git bash terminal inside VS Code and it wasn't working.

Switched the integrated terminal to cmd and everything started working great.

**Comment by NdozHH** [comment]
> > Had the same issue. Fixed by changing `remix.confix.js` **from**:
> > ```js
> > module.exports = {
> >   ignoredRouteFiles: [".*"],
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > **to**:
> > ```js
> > module.exports = {
> >   cacheDirectory: "./node_modules/.cache/remix",
> >   ignoredRouteFiles: [".*", "**/*.css", "**/*.test.{js,jsx,ts,tsx}"],
> > };
> > ```
> 
> This worked for me! Thanks @nilobarp After a server restart that reload was hot hot hot 🔥

This worked for me too!

**Comment by michaeldebetaz** [comment]
I just updated to 1.5.1 and still am experiencing this issue. Am I the only one?

```
WebSocket connection to 'ws://localhost:3001/socket' failed:

Remix dev asset server web socket error:

Event
 eventPhase: 2, …}
 isTrusted: true
 bubbles: false
 cancelBubble: false
 cancelable: false
 composed: false
 currentTarget: WebSocket {url: 'ws://localhost:3001/socket', readyState: 3, 
bufferedAmount: 0, onopen: null, onerror: ƒ, …}
 defaultPrevented: false
 eventPhase: 0
 path: []
 returnValue: true
 srcElement: WebSocket {url: 'ws://localhost:3001/socket', readyState: 3, 
 bufferedAmount: 0, onopen: null, onerror: ƒ, …}
 target: WebSocket {url: 'ws://localhost:3001/socket', readyState: 3, bufferedAmount: 0, onopen: null, onerror: ƒ, …}
 timeStamp: 2514.599999964237
 type: "error"
```

**Comment by tomadimitrie** [comment]
Same issue here, only in Docker it seems to not work
Running it on the host works though
(M1 Max)

**Comment by tomadimitrie** [comment]
Managed to fix this:
Because the live reload is just a WebSocket, it wasn't by default published in Docker
So I provided a fixed port in remix.config.js:
```
  devServerPort: 3001,
```
And published the port in my docker-compose.yml:
```
    ports:
      - "3001:3001"
```

Hope this helps someone

**Comment by emred2700** [comment]
Netlify Starter is not hot/live reloading. 

The OS doesn't matter. I tried in both macOS and Windows10. It doesn't show any errors on the terminal. The changes on files don't make the browser reload.

**Comment by kettanaito** [comment]
This looks like a bug in the `LiveReload` implementation:

https://github.com/remix-run/remix/blob/15213866f8dca94e6164c46028f410625226599f/packages/remix-react/components.tsx#L1367

You cannot reference `process` in a browser environment, it doesn't exist there. It's either those expressions are meant to be evaluated during the build, or they are simply incorrect. 

Copying a `LiveReload` into your setup is a bad idea. You'd be de-syncing your Remix setup from what the Remix team recommends (you won't get any fixes to `LiveReload`, for example).

**Comment by kettanaito** [comment]
What fixed the issue for me was setting an explicit `port` prop on `LiveReload`:

```tsx
import { LiveReload } from '@remix-run/react'

export default function App() {
  return (
    <html>
      <body><LiveReload port={8002} /></body>
    </html>
  )
}
```

It will stop attempting to look up the default value from the `process` then.

**Comment by sharuma23** [comment]
I had this issue on Windows. Some other user here (forgot the name sorry :P) said something about an issue with the WSL filesystem so I made the project in my home directory (~), opened it up with `code` and then the live reloads and whatever worked just fine. Also just to note, VSCode used the WSL remote thingy. 
![image](https://user-images.githubusercontent.com/71516284/173202907-633fefad-80f2-4bdc-8679-b4ef207e2ef2.png)

**Comment by shtwzrd** [comment]
Live reload does not seem to work for the socket.io example.

Setting `devServerPort` doesn't appear to make a difference in this case.

`remix watch` is seeing the file changes and rebuilding, and there is a websocket open on 8002. No errors in the browser console, and on initial page load we have successful connection to the websocket at 8002, but no network activity happens in response to the change, and reload requires manually refreshing the page.

**Comment by thatvegandev** [comment]
> Live reload does not seem to work for the socket.io example.

@shtwzrd I thought so too at first, but you just need to run the `dev` and `start:dev` scripts in parallel

**Comment by shtwzrd** [comment]
> > Live reload does not seem to work for the socket.io example.
> 
> @shtwzrd I thought so too at first, but you just need to run the `dev` and `start:dev` scripts in parallel

Hmm, I am doing exactly that.

But it is working for you, then? Did you make any changes, or is it a direct clone of main?

**Comment by edlaver** [comment]
In case it helps others, I tried a bunch of the stuff above before realising that Remix does "live reloads" and **not** "hot module replacement" aka HMR.

So live reloads **were** actually working (e.g. page was reloaded on changes automatically), but I was assuming that it should be doing HMR (e.g. changes updated in place without reloading the entire page).  This is not the case.

There's a good discussion here (https://github.com/remix-run/remix/discussions/2384) about HMR support, which will hopefully land eventually, although I appreciate the reasoning behind leaving it out currently.

**Comment by spencersmb** [comment]
> ```tsx
> port={8002}
> ```

This worked for me! I was having an issue where live reload worked just fine until I added or deleted a new file. Then I would get a websocket connection error after that and need to restart the remix server. This seems to have fixed it - great solution, thanks for posting.

**Comment by wlockiv** [comment]
> ```ts
> export const LiveReload =
>   process.env.NODE_ENV !== "development"
>     ? () => null
>     : function LiveReload({
>         port = Number(process.env.REMIX_DEV_SERVER_WS_PORT || 8002),
>       }: {
>         port?: number;
>       }) {
>         let setupLiveReload = ((port: number) => {
>           let protocol = location.protocol === "https:" ? "wss:" : "ws:";
>           let host = location.hostname;
>           let socketPath = `${protocol}//${host}:${port}/socket`;
> 
>           let ws = new WebSocket(socketPath);
>           ws.onmessage = (message) => {
>             let event = JSON.parse(message.data);
>             if (event.type === "LOG") {
>               console.log(event.message);
>             }
>             if (event.type === "RELOAD") {
>               console.log("💿 Reloading window ...");
>               window.location.reload();
>             }
>           };
>           ws.onerror = (error) => {
>             console.log("Remix dev asset server web socket error:");
>             console.error(error);
>           };
>         }).toString();
> 
>         return (
>           <script
>             suppressHydrationWarning
>             dangerouslySetInnerHTML={{
>               __html: `(${setupLiveReload})(${JSON.stringify(port)})`,
>             }}
>           />
>         );
>       };
> ```

This is the only thing here that's worked for me.

**Comment by Ali-Parandeh** [comment]
I know this is an easy fix but it made me move to Nextjs for my big project as Remix was not ready in my opinion yet for a prod app. I didn't want to hit issues like and having to dig deep to figure out how to fix them

**Comment by HuyAms** [comment]
The LiveRoad is super slow on my app. I have to wait 5-6 seconds, sometimes, forever. I haven't figured out what would be the reason. Currently, I have to kill the port and restart the server with `npm run dev` when I made some changes.

**Comment by philipswan** [comment]
I synced up a separate version of my project and found that live reload worked on the separate version. After comparing files, I discovered that in the original version, where live reload was not working, there was an extra file ".vscode/settings.json", and inside the file was the text
```
{
  "liveServer.settings.port": 5501
}  
```
When I deleted this text live reload started working again.
I think I might have tried to install a "Live Server" vscode extension  at some point. Perhaps the file was created at that time. Not sure. Hope this helps someone!

**Comment by tamm** [comment]
I found another way to run into an issue with `<LiveReload />` that led me here, `<LiveReload />` can prevent `<Link />` navigation if included when page is not hydrated.

## How to replicate

Create a remix project and add the logic to trigger hydration only on some routes from [Disabling JavaScript](https://remix.run/docs/en/v1/guides/disabling-javascript)

Create a route `routes/index.tsx` where you include a `<Link to={hydrating-route}>Hydrating route</Link>`, and make sure this route does not hydrate.
And a second route like `routes/hydrating-route.tsx` where you include `export const handle = { hydrate: true };`

Now start the dev server and go to the first route, and click the link. You'll most likely end up on the destination link, if you're in Chrome. In Firefox 106.0.3 however (Windows and mac tested) I end up right where I started! The route appears to just bounce, which isn't exactly what happens but not far away either.

## What seems to be the problem in this case

As of the current release 1.7.4 of [@remix-run/react](https://github.com/remix-run/remix/blob/%40remix-run/react%401.7.4/packages/remix-react/components.tsx#L1538) the `<LiveReload />` component will automatically reload your page every time the websocket connection is reset, regardless of why it is reset. ([It's a feature](https://github.com/remix-run/remix/commit/2d46b913362dcb6c3c69a9fc46fba316565e00de))

So what happens in your (Firefox) browser when this goes wrong?
In short, the browser will:
* navigate to the linked resource
* websocket connection is broken
* new route takes long enough to respond that websocket is reconnected 
* window reload triggers
* your request to fetch the new route is aborted by the browser
* you see / again

How can you avoid this?

1. You can make sure you **always hydrate if `<LiveReload />` is going to be on**, perhaps by
    ```javascript
          const includeScripts = process.env.NODE_ENV === "development" || matches.some(
            (match) => match.handle?.hydrate
          );
    ```
2. You could instead make sure you **skip `<LiveReload />` when you are skipping hydration**.
    ```javascript
        {includeScripts ? <Scripts /> : null}
    ```
3. Replacing `<LiveReload />` with the inline version above works because it is from before the feature to reload on re-connect was added. This is probably a bad idea, but that's WHY that works.

## Why am I not reporting this as a separate bug instead?

I'm not sure it IS a bug, it might be more of a user error, and I intend to simply ignore it, as most of the time you do not need to worry about this edge-case.

**Comment by Ali-Parandeh** [comment]
This pretty much stopped me from using remix as my framework.

**Comment by machour** [comment]
@Ali-Parandeh You already mentioned that, I'm not sure repeating it is helpful :)

**Comment by Ali-Parandeh** [comment]
My apologies.

**Comment by pcattori** [comment]
We've made improvements to live reload (including fixing bugs for Firefox) and have a new dev server (available under the `unstable_dev` future flag) slated for the next minor release.

Many symptoms of buggy live reload have been raised here, but it's unclear if any of them are still an issue. I'm going to close the issue, but feel free to reopen if still having problems with it in Remix v1.11.1+

**Comment by philipswan** [comment]
I appreciate your efforts to make improvements! I think that it's not a good idea to close a bug simply because you've fixed other, potentially unrelated, bugs. This is especially true when the bug you are closing is serious and deserves a very focused and dedicated effort to: root cause it, fix it, and to create tests that are designed to prevent it from happening again. I would prefer it to be left open until its dealt with properly. If you would like the community's help to try out the new build to see if it makes some headway, please provide really detailed and clear instructions on how to do this - keeping in mind that people who are experiencing this problem will have no idea how to temporarily switch to, and then return back from, the new unstable build that you created.

**Comment by philipswan** [comment]
If you have 10 minutes, you could try syncing [this project](https://github.com/philipswan/TetheredRing) (also see: https://youtu.be/k5fki-C3r70) to see if you can reproduce the issue when using Chrome and when making changes to transitsystems.js. If you can, then try the new build to see if it fixes the issue. If you can't repo, then please send your contact info to me at contact@project-atlantis.com. Thanks!

**Comment by pcattori** [comment]
> I appreciate your efforts to make improvements! I think that it's not a good idea to close a bug simply because you've fixed other, potentially unrelated, bugs. This is especially true when the bug you are closing is serious and deserves a very focused and dedicated effort to: root cause it, fix it, and to create tests that are designed to prevent it from happening again. I would prefer it to be left open until its dealt with properly. If you would like the community's help to try out the new build to see if it makes some headway, please provide really detailed and clear instructions on how to do this - keeping in mind that people who are experiencing this problem will have no idea how to temporarily switch to, and then return back from, the new unstable build that you created.

To be clear, we've made improvements to the existing dev server _as well_ as introduced a new dev server. Also, the original issue is no longer reproducible and works as expected.

If anyone is still experiencing issues with recent Remix versions (say after Remix 1.9 or 1.11) then, I'd be happy for them to reopen this.

**Comment by pcattori** [comment]
> If you have 10 minutes, you could try syncing [this project](https://github.com/philipswan/TetheredRing) (also see: https://youtu.be/k5fki-C3r70) to see if you can reproduce the issue when using Chrome and when making changes to transitsystems.js. If you can, then try the new build to see if it fixes the issue. If you can't repo, then please send your contact info to me at [contact@project-atlantis.com](mailto:contact@project-atlantis.com). Thanks!

^this looks like its using Vite, not Remix

**Comment by philipswan** [comment]
Ahh, sorry, I am using node, but not so familiar with what's under the hood after that (e.g. Vite versus Remix). Is it easy to switch a project from Vite to Remix and back?

**Comment by microtactical** [comment]
LiveReload is still not working in version 1.13. If you include the <Scripts/> import from "@remix-run/react", then it works.

**Comment by pcattori** [comment]
> LiveReload is still not working in version 1.13. If you include the `<Scripts/>` import from "@remix-run/react", then it works.

v1.13 isn't released yet. Are you referring to v1.12? Or a `nightly` build?

**Comment by microtactical** [comment]
> 

Apologies, yes 1.12. I had NextJS 13 in my head as I'm trying different SSR frameworks out!. Yes, 1.12, but the rest of what I stated still stands.

**Comment by pcattori** [comment]
> > 
> 
> Apologies, yes 1.12. I had NextJS 13 in my head as I'm trying different SSR frameworks out!. Yes, 1.12, but the rest of what I stated still stands.

Should be fixed by #5339

**Comment by observerkei** [comment]
If you are using wsl to view windows files, please move the project inside the wsl

**Comment by dreymoreau** [comment]
> @LukeXF Are you on Mac or Windows?
> 
> Did you add/change `devServerPort` in remix.config?
> 
> Also check to see if `REMIX_DEV_SERVER_WS_PORT` environment variable is defined.
> 
> Worst case, just update your `dev` script to:
> 
> `"cross-env NODE_ENV=development PORT=3000 REMIX_DEV_SERVER_WS_PORT=8002 remix dev"`
> 
> Also, try to connect to the web socket server using your browser: http://localhost:8002
> 
> You should see the message "Upgrade Required"



> @LukeXF Are you on Mac or Windows?
> 
> Did you add/change `devServerPort` in remix.config?
> 
> Also check to see if `REMIX_DEV_SERVER_WS_PORT` environment variable is defined.
> 
> Worst case, just update your `dev` script to:
> 
> `"cross-env NODE_ENV=development PORT=3000 REMIX_DEV_SERVER_WS_PORT=8002 remix dev"`
> 
> Also, try to connect to the web socket server using your browser: http://localhost:8002
> 
> You should see the message "Upgrade Required"

this helped!!! thanks so much :D

**Comment by KilloconQ** [comment]
Hi I hace the same problem, I am using the new version remix 2.0.1 and live reload is not working.

**Comment by rodolphoasb** [comment]
If anyone tried everything that was suggested here and it still didn't work, I just restarted my MacBook and it's working now, seriously.

**Comment by BoilingSoup** [comment]
The official remix tutorial runs straight into this problem. Not a good first experience with the framework

**Comment by adaboese** [comment]
What's the latest here?

**Comment by karolis-sh** [comment]
I had previously removed both `app/entry.server.tsx` and `app/entry.client.tsx` (as I thought they were redundant based on generated file comments), regenerating them via `npx remix reveal` seems to have fixed issues wit live reloading

**Comment by enBonnet** [comment]
I fix this by adding the `components` directory that I have outside of the `app` directory in the `watchPaths` array, my `remix.conf.js` file looks like this:

```js
/** @type {import('@remix-run/dev').AppConfig} */
export default {
  ignoredRouteFiles: ["**/.*"],
  watchPaths: ["./components"],
  // appDirectory: "app",
  // assetsBuildDirectory: "public/build",
  // publicPath: "/build/",
  // serverBuildPath: "build/index.js",
};

```

**Comment by rodolphoasb** [comment]
You can now use Vite, which should improve this 🥳 https://remix.run/docs/en/main/future/vite#migrating

**Comment by camdowney** [comment]
Guess I should've actually read the docs first... fixed everything by changing my root.tsx from this:
`export default () => {`

To this:
`export default function Root() {`

I have no idea if this is a common limitation across frameworks, but the docs clearly state not to do what I was doing: https://remix.run/docs/en/main/discussion/hot-module-replacement.

**Comment by enBonnet** [comment]
> Guess I should've actually read the docs first... fixed everything by changing my root.tsx from this:
> 
> ```js
> export default () => {
> ```
> 
> To this:
> 
> ```js
> export default function Root() {
> ```
> 
> I have no idea if this is a common limitation across frameworks, but the docs clearly state not to do what I was doing: https://remix.run/docs/en/main/discussion/hot-module-replacement.

Good to know.

**Comment by adaboese** [comment]
Good to know!

**Comment by atalayozmen** [comment]
I was having the same issue with the blues stack, I started with the template that had <LiveReload/> after <Scripts/>. I changed the order now where <LiveReload/> comes first, and now it is fixed. This is important if you are using vitest.
Source: https://remix.run/docs/en/main/future/vite#livereload-before-scripts
`     
        <body className="h-full">
          <Outlet />
          <ScrollRestoration />
          **<LiveReload />**
          **<Scripts />**
      </body>
`

**Comment by alisnino** [comment]
This is the most basic feature a frontend dev framework should have; it's not working in Nov/2023 so I'll report to my team Remix is not worth considering yet 😇

**Comment by enBonnet** [comment]
> This is the most basic feature a frontend dev framework should have; it's not working in Nov/2023 so I'll report to my team Remix is not worth considering yet 😇

It is working, but we have to improve the documentation because it'd work properly if you keep your code inside the `/app` directory.

**Comment by adaboese** [comment]
My structure is as follows:

```
└── src
    ├── app
    │   ├── defer
    │   └── routes
    │       ├── ...
    ├── components
    │   ├── previews
    │   │   └── data-table
    │   └── ui
    │       └── toast
    ├── hooks
    ├── routines
    ├── services
    ├── styled-system
    │   ├── css
    │   ├── jsx
    │   ├── patterns
    │   ├── recipes
    │   ├── tokens
    │   └── types
    └── utilities
```

Does this mean I need to move `src/hooks/`, `src/components/`, ... to `src/app/`?

Do I need to use named function for everything?

**Comment by pcattori** [comment]
There seem to be multiple different issues people are running into.

## Named exports

[Named exports are required for HMR](https://remix.run/docs/en/main/discussion/hot-module-replacement#named-function-components). Note that this is a limitation of React Fast Refresh, not Remix. Thus the same limitation is present in Vite and [Next](https://nextjs.org/docs/architecture/fast-refresh#limitations).

## Code outside of `app/` dir

Expecting changes outside of `app/` to trigger hot updates. As already mentioned, you can add those paths to `watchPaths`, but we'd recommend keeping all your app code within `app/` anyway.

## `LiveReload` before `Scripts` (Vite)

This limitation has been removed from unstable Vite support. The order of `LiveReload` and `Scripts` no longer matters.

## Quickstart/Tutorial not working

@BoilingSoup I couldn't reproduce this error, but if you are still running into it could you file a separate issue?

---

This issue was already too broad when it was first created, so moving forward please file separate issues with your specific problem so that we can make sure we prioritize a fix.

**Comment by pcattori** [comment]
@alisnino are you facing a different issue than the ones I mentioned?

**Comment by rodolphoasb** [comment]
Hey, @pcattori, first I'd like to say thank you for your amazing work integrating Vite.

Here's an issue that I opened related to the Vite live reload: https://github.com/remix-run/remix/issues/7853 (there's a repo where you can reproduce it), and tbh it makes the DX really bad. 

Not sure how complex it is, but happy to take a look at the weekend if you could give me some pointes...

**Comment by pcattori** [comment]
@rodolphoasb let's keep the discussion for that on your issue

**Comment by adaboese** [comment]
Can anyone recommend a workaround for naming functions? Using arrow functions would have quite a few downsides

**Comment by BoilingSoup** [comment]
> @BoilingSoup I couldn't reproduce this error, but if you are still running into it could you file a separate issue?

https://github.com/remix-run/remix/issues/7466

It is here.

**Comment by pcattori** [comment]
@adaboese you can use `function` keyword or arrow functions as long as they are named. [Examples of this are in the docs](https://remix.run/docs/en/main/discussion/hot-module-replacement#named-function-components).

**Comment by dharmendra94** [comment]
Hey @pcattori, Even in the named exported functions from a package outside the app directory as below: 

```
In repo/packages/home

export function Home() {
  return (<>Content here</>);
}

In repo/apps/remix-app/routes/home.tsx

import { Home, handle } from "@scope/home-package";
import { action, loader } "@scope/home-package/server";

export default Home;

export { action, loader, handle };

```

I am observing the HMR is not working when I make a change inside the Home package (I got it under `watchPaths`, and the connection from the dev server to the browser's `LiveReload` is working fine.

================

The moment I change the route file as below, HMR starts working.

```
In repo/apps/remix-app/routes/home"

import { Home } from '@scope/home-package';

export default function HomePage() {
  return Home();
}
```

**Comment by dharmendra94** [comment]
Hey, @pcattori, I have tried Vite, and the issue no longer exists. Now I am curious, when will `Vite` be stable and ready for production? 🤔

**Comment by adaboese** [comment]
> Code outside of app/ dir
> Expecting changes outside of app/ to trigger hot updates. As already mentioned, you can add those paths to watchPaths, but we'd recommend keeping all your app code within app/ anyway.

I think `watchPaths` does not exist as an option when using with Vite plugin?

**Comment by kiliman** [comment]
@adaboese 

> I think `watchPaths` does not exist as an option when using with Vite plugin? 

I haven't tried it yet, but you might try adding `esbuild.include` to your _vite.config.js_ to add additional paths. 

https://vitejs.dev/config/shared-options.html#esbuild

**Comment by pcattori** [comment]
> Hey, @pcattori, I have tried Vite, and the issue no longer exists. Now I am curious, when will `Vite` be stable and ready for production? 🤔

https://twitter.com/markdalgleish/status/1736240852308251053

**Comment by Chinwike1** [comment]
2024 and this issue still exists. Its gotten so frustrating

**Comment by pcattori** [comment]
> 2024 and this issue still exists. Its gotten so frustrating

@Chinwike1 this discussion has been closed because its been resolved with our move to Vite. If its not working for you and you have a reproduction (ideally on StackBlitz), you can open a separate GitHub issue so that we can fix it.

**Comment by willlogs** [comment]
@Chinwike1 Exactly...
I love Remix but this happens randomly a lot. Sometimes the hot reloading breaks and after a few days it starts working again. It's really random.

**Comment by ttebify** [comment]
It is 2024.

I am adding this comment because I came here after experiencing the same issue.

If you are using TailwindCSS, make sure your stylesheet is imported correctly. It appears this step might be omitted in the default `create-remix` app setup.

Here are the changes I made to make MHR work:

1. Import your stylesheet like this:  
   ```tsx
   import stylesheet from "~/tailwind.css?url";
   ```

2. Add it to the `links` export:  
   ```tsx
   export const links: LinksFunction = () => [
     { rel: "stylesheet", href: stylesheet },
     // ... Other links
   ];
   ```

It’s as simple as following the instructions in the [TailwindCSS documentation](https://tailwindcss.com/docs/guides/remix).

**Comment by nickfthedev** [comment]
> > I still have not found a solution other than it the hot-reloading it working on my Linux machine but refuses to work on Windows.
> 
> I hit the same problems, then realised that WSL was sat in the middle. Live Reload started working perfectly once I moved my working directory to the Linux filesystem. https://docs.microsoft.com/en-us/windows/wsl/setup/environment#file-storage

You made my day! I accedentially added the project into the wrong /dev folder

**Comment by erfurtjohn** [comment]
> It is 2024.
> 
> I am adding this comment because I came here after experiencing the same issue.
> 
> If you are using TailwindCSS, make sure your stylesheet is imported correctly. It appears this step might be omitted in the default `create-remix` app setup.
> 
> Here are the changes I made to make MHR work:
> 
> 1. Import your stylesheet like this:
>    import stylesheet from "~/tailwind.css?url";
> 2. Add it to the `links` export:
>    export const links: LinksFunction = () => [
>      { rel: "stylesheet", href: stylesheet },
>      // ... Other links
>    ];
> 
> It’s as simple as following the instructions in the [TailwindCSS documentation](https://tailwindcss.com/docs/guides/remix).

This helped me out.
